### PR TITLE
Adding public_filters to SearchFacet for anonymous searches (SCP-3471)

### PR DESCRIPTION
**Pull request by bistline**
Currently, when viewing the SCP home page before logging in, the filter values displayed when clicking on a search facet can contain values that only reside in private studies that the user may not be able to view, once logged in.  This can lead to confusing situations where a user selects one of those filter values, runs a search, and gets no results (due to correct permission filtering on studies).

This update adds the `public_filters` attribute to `SearchFacet`, where only filter values from public studies are shown.  This way, when an anonymous user selects a filter value, they are guaranteed to see at least one result.  Once logged in, the former scenario still applies, but solving that issue is out of scope for this PR.  In addition, this update applies general linting format updates to files updated in this PR.

MANUAL TESTING
Since waiting for an ingest of a convention-compliant metadata file in a private study will take several minutes, the easiest way to test this feature is to manually update the `public_filters` value on a search facet.  Also, you will need to set the default value for the "faceted_search" `FeatureFlag` to `true` in order to have advanced search show up when not logged in.

1. From the project root, after running `rails_local_setup.rb`, run all queued migrations:
```
bin/rails db:migrate
```
2. Enter the rails console, and find the 'faceted_search' `FeatureFlag`:
```
feature_flag = FeatureFlag.find_by(name: 'faceted_search')
```
3. Set the `default_value` to true:
```
feature_flag.update(default_value: true)
```
4. Load the `SearchFacet` for 'species'
```
facet = SearchFacet.find_by(name: 'species')
```
5. Update the `public_filters` to be only the first two filters:
```
facet.update(public_filters: facet.filters.take(2))
```
6. In another terminal window, boot the portal as normal, and click on the "species" facet, confirming there are only 2 values:
Screen Shot 2021-07-12 at 3 14 49 PM
7. Sign in, and confirm that all values for species are now shown:
Screen Shot 2021-07-12 at 3 15 42 PM
8. (Optional): Back in the rails console, reset the 'species' facet:
```
facet.update_filter_values!
```

This PR satisfies SCP-3471

### CODE REVIEW

File: app/models/search_facet.rb [hunk 1]
```diff
@@ -164,10 +166,10 @@ class SearchFacet
     end
     property :filters do
       key :type, :array
-      key :description, 'Array of filter values for facet'
+      key :description, 'Array of filter values for facet (will default to public_filters if no user is signed in)'
```
Comment (eweitz) [inline app/models/search_facet.rb:169]: Blocker: there seems to be a subtle documentation bug that substantially changes meaning.

```suggestion
      key :description, 'Array of filter values for facet (will default to public_filters if user is not signed in)'
```

File: app/models/search_facet.rb [hunk 2]
```diff
@@ -309,80 +311,96 @@ def convert_time_between_units(base_value:, original_unit:, new_unit:)
...
-    values = self.get_unique_filter_values
-    unless values.empty?
-      if self.is_numeric?
-        self.update(min: values[:MIN], max: values[:MAX])
-      else
-        self.update(filters: values.to_a)
-      end
+    if is_numeric?
+      values = get_unique_filter_values
+      return false if values.empty? # found no results, meaning an error occurred
```
Comment (eweitz) [inline app/models/search_facet.rb:349]: If an error occurred, should it be thrown and logged here?  I'd think so, but might be overlooking something.

**Review by eweitz — CHANGES_REQUESTED** [review]
This will soften a search UX papercut!

I noted a trivial blocker, and suggested a non-blocking maintainability improvement.

Although this fix for unauth'd users doesn't fix how auth'd users see some filters for which no results will be returned, does this work lower effort for fixing that?  Or would fixing that related issue require an orthogonal solution?

File: app/models/search_facet.rb (hunk 2 above)
Comment (bistline) [reply app/models/search_facet.rb:349]: That is handled in `get_unique_filter_values`, which is where the actual call to BigQuery happens.  This is just the setter.

**Review by eweitz — APPROVED** [review]

**Comment by bistline** [comment]
> Although this fix for unauth'd users doesn't fix how auth'd users see some filters for which no results will be returned, does this work lower effort for fixing that? Or would fixing that related issue require an orthogonal solution?

Not really... the only way I can see of truly solving this problem completely is maintaining a map of all facet/filter values to study accessions, and then filtering the UI based on which studies the user is allowed to see.  That seems a rather complicated and expensive solution to an admittedly minor UX grumble.

**Review by ehanna4 — APPROVED** [review]
Looks good, tested manually following the repro steps provided and compared to dev and this definitely is a nicer UX. Code looks good as far as I can tell. The use of `%w[]` for string arrays is a neat ruby-feature/change

**Pull request merged**